File: packages/typescript/src/features/advanced-memory/actions/resetSession.ts
/**
 * Reset Session Action
 *
 * Resets the conversation session by setting a compaction point.
 * Messages before the compaction point will not be loaded in future context.
 *
 * Only OWNER or ADMIN roles can execute this action.
 */

import { logger } from "../../../logger.ts";
import { getUserServerRole } from "../../../roles.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";

export const resetSessionAction: Action = {
	name: "RESET_SESSION",
	similes: ["CLEAR_HISTORY", "NEW_SESSION", "FORGET", "START_OVER", "RESET"],
	description:
		"Resets the conversation session by creating a compaction point. Messages before this point will not be included in future context. Use when the user wants to start fresh or clear conversation history.",

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
	): Promise<boolean> => {
		if (!state) {
			return false;
		}

		const room = state.data.room ?? (await runtime.getRoom(message.roomId));
		if (!room?.worldId) {
			// Allow in DMs without world/server context
			return true;
		}

		// Check user has permission
		const userRole = await getUserServerRole(
			runtime,
			message.entityId,
			room.worldId,
		);

		return userRole === "OWNER" || userRole === "ADMIN";
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
		_responses?: Memory[],
	): Promise<ActionResult> => {
		try {
			const room = state?.data.room ?? (await runtime.getRoom(message.roomId));

			if (!room) {
				logger.error(
					{ src: "action:reset-session", roomId: message.roomId },
					"Room not found",
				);
				if (callback) {
					await callback({
						text: "Unable to reset session - room not found.",
						actions: ["RESET_SESSION_FAILED"],
						source: message.content.source,
					});
				}
				return {
					text: "Room not found",
					success: false,
					values: { error: "room_not_found" },
					data: { actionName: "RESET_SESSION" },
				};
			}

			const now = Date.now();
			const previousCompaction = room.metadata?.lastCompactionAt;

			// Update room metadata with compaction point
			await runtime.updateRoom({
				...room,
				metadata: {
					...room.metadata,
					lastCompactionAt: now,
					compactionHistory: [
						...(Array.isArray(room.metadata?.compactionHistory)
							? room.metadata.compactionHistory
							: []),
						{
							timestamp: now,
							triggeredBy: message.entityId,
							reason: "manual_reset",
						},
					].slice(-10), // Keep last 10 compaction events
				},
			});

			logger.info(
				{
					src: "action:reset-session",
					roomId: message.roomId,
					entityId: message.entityId,
					compactionAt: now,
				},
				"Session reset - compaction point set",
			);

			if (callback) {
				await callback({
					text: "Session has been reset. I'll start fresh from here.",
					actions: ["RESET_SESSION"],
					source: message.content.source,
				});
			}

			return {
				text: "Session reset successfully",
				success: true,
				values: {
					success: true,
					compactionAt: now,
					previousCompactionAt: previousCompaction,
					roomId: room.id,
				},
				data: {
					actionName: "RESET_SESSION",
					compactionAt: now,
					roomId: room.id,
				},
			};
		} catch (error) {
			logger.error(
				{ src: "action:reset-session", error },
				"Error resetting session",
			);

			if (callback) {
				await callback({
					text: "Sorry, I encountered an error while trying to reset the session.",
					actions: ["RESET_SESSION_FAILED"],
					source: message.content.source,
				});
			}

			return {
				text: `Error resetting session: ${error instanceof Error ? error.message : "Unknown error"}`,
				success: false,
				values: { error: String(error) },
				data: { actionName: "RESET_SESSION" },
			};
		}
	},

	examples: [
		[
			{
				name: "{{name1}}",
				content: {
					text: "Let's start over. Reset the session.",
				},
			},
			{
				name: "{{agentName}}",
				content: {
					text: "Session has been reset. I'll start fresh from here.",
					actions: ["RESET_SESSION"],
				},
			},
		],
		[
			{
				name: "{{name1}}",
				content: {
					text: "Clear our conversation history",
				},
			},
			{
				name: "{{agentName}}",
				content: {
					text: "Session has been reset. I'll start fresh from here.",
					actions: ["RESET_SESSION"],
				},
			},
		],
		[
			{
				name: "{{name1}}",
				content: {
					text: "Forget everything we talked about",
				},
			},
			{
				name: "{{agentName}}",
				content: {
					text: "Session has been reset. I'll start fresh from here.",
					actions: ["RESET_SESSION"],
				},
			},
		],
	] as ActionExample[][],
};

File: packages/python/elizaos/features/advanced_memory/actions/reset_session.py
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class ResetSessionAction:
    name: str = "RESET_SESSION"
    similes: list[str] = field(
        default_factory=lambda: ["CLEAR_HISTORY", "NEW_SESSION", "FORGET", "START_OVER", "RESET"]
    )
    description: str = (
        "Resets the conversation session by creating a compaction point. "
        "Messages before this point will not be included in future context."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, state: State | None = None
    ) -> bool:
        room = None
        if state and getattr(state, "data", None) and getattr(state.data, "room", None):
            room = state.data.room
        elif message.room_id:
            room = await runtime.get_room(message.room_id)

        if not room or not getattr(room, "world_id", None):
            return True

        world = await runtime.get_world(room.world_id)
        if not world or not getattr(world, "metadata", None):
            return False

        roles = world.metadata.get("roles", {})
        user_role = roles.get(str(message.entity_id), "NONE")
        return user_role in ("OWNER", "ADMIN")

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        del options, responses

        room = None
        if state and getattr(state, "data", None) and getattr(state.data, "room", None):
            room = state.data.room
        elif message.room_id:
            room = await runtime.get_room(message.room_id)

        if room is None:
            if callback:
                await callback(
                    Content(
                        text="Unable to reset session - room not found.",
                        actions=["RESET_SESSION_FAILED"],
                        source=getattr(message.content, "source", None),
                    )
                )
            return ActionResult(
                text="Room not found",
                values={"error": "room_not_found"},
                data={"actionName": "RESET_SESSION"},
                success=False,
            )

        metadata = dict(getattr(room, "metadata", {}) or {})
        previous_compaction = metadata.get("lastCompactionAt")
        compaction_history = list(metadata.get("compactionHistory", []))
        now = int(time.time() * 1000)

        compaction_history.append(
            {
                "timestamp": now,
                "triggeredBy": str(message.entity_id),
                "reason": "manual_reset",
            }
        )
        metadata["lastCompactionAt"] = now
        metadata["compactionHistory"] = compaction_history[-10:]

        if (
            hasattr(room, "metadata")
            and hasattr(room.metadata, "clear")
            and hasattr(room.metadata, "update")
        ):
            room.metadata.clear()
            room.metadata.update(metadata)
        else:
            room.metadata = metadata

        await runtime.update_room(room)

        if callback:
            await callback(
                Content(
                    text="Session has been reset. I'll start fresh from here.",
                    actions=["RESET_SESSION"],
                    source=getattr(message.content, "source", None),
                )
            )

        result = ActionResult(
            text="Session reset successfully",
            values={
                "success": True,
                "compactionAt": now,
                "roomId": str(room.id),
            },
            data={
                "actionName": "RESET_SESSION",
                "compactionAt": now,
                "roomId": str(room.id),
            },
            success=True,
        )
        if previous_compaction is not None:
            result.values["previousCompactionAt"] = previous_compaction
        return result

    @property
    def examples(self) -> list[list[ActionExample]]:
        return []


reset_session_action = Action(
    name=ResetSessionAction.name,
    similes=ResetSessionAction().similes,
    description=ResetSessionAction.description,
    validate=ResetSessionAction().validate,
    handler=ResetSessionAction().handler,
    examples=ResetSessionAction().examples,
)
